### PR TITLE
Update solvers to handle binomial.

### LINKED ISSUE #10993: Solving equations with binomial coefficients should expand them first
When trying to solve an equation with a binomial coefficient with a symbol, the solver fails:

``` python
>>> solve(Eq(binomial(x, 3), 1000))
```

``` pytb
Traceback (most recent call last):
  File "<ipython-input-7-e56e9952e4b8>", line 1, in <module>
    solve(Eq(binomial(x, 3), 1000))
  File "sympy/solvers/solvers.py", line 1053, in solve
    solution = _solve(f[0], *symbols, **flags)
  File "sympy/solvers/solvers.py", line 1619, in _solve
    raise NotImplementedError('\n'.join([msg, not_impl_msg % f]))
NotImplementedError: 
No algorithms are implemented to solve equation binomial(x, 3) - 1000
```

But when expanding the binomial coefficient first, the solver succeeds:

``` python
>>> solve(Eq(binomial(x, 3).expand(func=True), 1000))[2].n()
19.1895499623678
```

Shouldn't the solver try to expand the binomial before failing?

### PR DESCRIPTION
**Pull request by sbt4104**
#### References to other Issues or PRs
Fixes #10993

#### Brief description of what is fixed or changed
Added conditions to expand binomial in solvers
Update expand function for binomial
#### Other comments 
Please suggest changes.

### Release Notes

* solvers
    - Fix solvers to handle binomials
* functions
    - Update expand function for binomials.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
I'm not sure where this belongs but certainly not as the opening lines in the solve function. Unfortunately solve is a bit of a mess so there isn't necessarily a correct place to put something like this.

**Comment by sbt4104** [comment]
I think putting those conditions in "preprocess equations" is a more right.

**Comment by sbt4104** [comment]
Please review and suggest the changes.

**Comment by oscarbenjamin** [comment]
Consider that you might have an expression involving a binomial like
```
In [4]: expr = 1 + binomial(x, 3)                                                                                                                             

In [5]: isinstance(expr, binomial)                                                                                                                            
Out[5]: False

In [6]: expr.has(binomial)                                                                                                                                    
Out[6]: True
```
We probably don't want to expand all functions so it would be better to use expr.replace to pick out the binomials and expand them.

This code is still not in the right place: solve hasn't even converted Eqs to Expr at that point. Look through solve to find where similar transformations are happening.

**Comment by sbt4104** [comment]
I have made the changes so that only binomial terms are expanded and added the code after Eqs are checked and converted. Please check if it is correctly done.

**Comment by sbt4104** [comment]
can someone explain to me why my test on travis-ci is failing.
`$ bin/test_travis.sh`
` +[[ '' == \t\r\u\e ]] `
 `+[[ '' == \t\r\u\e ]] `
 `+[[ '' == \t\r\u\e ]] `
 `+[[ '' == \t\r`

**Comment by sbt4104** [comment]
I pushed the same changes again and it passed.
Please review the changes.

File: sympy/solvers/solvers.py [hunk 1]
```diff
@@ -999,6 +999,9 @@ def _sympified_list(w):
                     fi = fi.rewrite(Add, evaluate=False)
             f[i] = fi
 
+        if fi.has(binomial):
+            f[i] = fi.replace(lambda exp: binomial, lambda exp: expand_func(exp))
```
Comment (oscarbenjamin) [inline sympy/solvers/solvers.py]: The filter function here will return true for any expression:
```
In [1]: f = lambda exp: binomial                                                                                                               

In [2]: f(sin(x))                                                                                                                              
Out[2]: binomial

In [3]: bool(f(sin(x)))                                                                                                                        
Out[3]: True
```

Comment (sbt4104) [reply sympy/solvers/solvers.py]: I have made the change for it to correctly identify the binomials.

**Comment by oscarbenjamin** [comment]
I think translating binomials should still come later after the loop that it is in now.

This will need to have a test added for equations with binomial coefficients that can be solved.

**Comment by sbt4104** [comment]
I will make the changes.

**Comment by sbt4104** [comment]
I have added the test.

**Comment by sbt4104** [comment]
I have added the binomial expansion after the loop, please suggest changes.

File: sympy/solvers/solvers.py [hunk 2]
```diff
@@ -1027,6 +1027,11 @@ def _sympified_list(w):
                     bare_f = False
                 f[i: i + 1] = [fr, fi]
 
+    #expand binomial terms
+    for i, fi in enumerate(f):
+        if fi.has(binomial):
+            f[i] = fi.replace(lambda exp: binomial, lambda exp: expand_func(exp))
```
Comment (oscarbenjamin) [inline sympy/solvers/solvers.py]: The filter function here will still match non-binomial expressions.

Comment (sbt4104) [reply sympy/solvers/solvers.py]: I will correct it right away.

Comment (oscarbenjamin) [reply sympy/solvers/solvers.py]: How about:
```python
expand_binomial = lambda f: f.replace(binomial, lambda n, k: expand_func(binomial(n, k)))
f = [expand_binomial(fi) for fi in f]
```

Comment (sbt4104) [reply sympy/solvers/solvers.py]: This one looks more efficient, I will make the changes.

Comment (sbt4104) [reply sympy/solvers/solvers.py]: I earlier thought that binomial expansion should be done in the previous loop as hyperbolic functions are checked and rewritten inside it.

Comment (sbt4104) [reply sympy/solvers/solvers.py]: please review the changes.

**Comment by oscarbenjamin** [comment]
On master we have:
```julia
In [1]: solve(x-binomial(y, 10))                                                                                                                              
Out[1]: 
⎡⎧   ⎛y ⎞⎫⎤
⎢⎨x: ⎜  ⎟⎬⎥
⎣⎩   ⎝10⎠⎭⎦
```
This PR gives
```julia
In [1]: solve(x-binomial(y, 10))                                                                                                                              
Out[1]: 
⎡⎧     ⎛ 9       8        7         6          5           4           3            2                     ⎞⎫⎤
⎢⎪   y⋅⎝y  - 45⋅y  + 870⋅y  - 9450⋅y  + 63273⋅y  - 269325⋅y  + 723680⋅y  - 1172700⋅y  + 1026576⋅y - 362880⎠⎪⎥
⎢⎨x: ──────────────────────────────────────────────────────────────────────────────────────────────────────⎬⎥
⎢⎪                                                  3628800                                                ⎪⎥
⎣⎩                                                                                                         ⎭⎦
```
I think the behaviour on master is better so I guess the replace call needs to be modified so that it only expands binomials that involve the unknowns that we are trying to solve for.

**Comment by sbt4104** [comment]
I have made the changes, please check.

**Comment by sbt4104** [comment]
It is failing if in binomial(n,k) n is a polynomial in x, example
`y = solve(1-binomial(x+1, 3),x)`
I think adding this will remove this error
`f = f.replace(binomial, lambda n, k: expand_func(binomial(n, k)) if symbol in n.free_symbols and len(n.free_symbols) == 1 else binomial(n,k))`

**Comment by sbt4104** [comment]
@oscarbenjamin should I make this change?

**Comment by oscarbenjamin** [comment]
> It is failing if in binomial(n,k) n is a polynomial in x, example
> `y = solve(1-binomial(x+1, 3),x)`

In what way was it failing?

**Comment by sbt4104** [comment]
I had written it like this in the previous commit
`f = f.replace(binomial, lambda n, k: expand_func(binomial(n, k)) if n==symbol else binomial(n, k))`
so for `y = solve(1-binomial(x+1, 3),x)`, it was comparing x+1 with x and returned 'NotImplementedError: 
No algorithms are implemented to solve equation binomial(x + 3, 3)'

 though it worked perfectly for  `y = solve(1-binomial(x+1, 3), x+1)`


so i changed it to solve cases like this 
`f = f.replace(binomial, lambda n, k: expand_func(binomial(n, k)) if symbol in n.free_symbols and len(n.free_symbols) == 1 else binomial(n,k))`

**Comment by oscarbenjamin** [comment]
This is probably best:
```
In [5]: syms = [x]                                                                                                                                            

In [6]: binomial(x, 2).replace(lambda e: isinstance(e, binomial) and e.has(*syms), lambda e: expand_func(e))                                                  
Out[6]: 
x⋅(x - 1)
─────────
    2  
```

**Comment by oscarbenjamin** [comment]
It looks like `expand_func` for binomial could be improved:
```
In [18]: expand_func(binomial(x, 2))                                                                                                                          
Out[18]: 
x⋅(x - 1)
─────────
    2    

In [19]: expand_func(binomial(x+2, x))                                                                                                                        
Out[19]: 
⎛x + 2⎞
⎜     ⎟
⎝  x  ⎠

In [20]: gammasimp(_.rewrite(factorial))                                                                                                                      
Out[20]: 
(x + 1)⋅(x + 2)
───────────────
       2   
```

**Comment by sbt4104** [comment]
> This is probably best:
> 
> ```
> In [5]: syms = [x]                                                                                                                                            
> 
> In [6]: binomial(x, 2).replace(lambda e: isinstance(e, binomial) and e.has(*syms), lambda e: expand_func(e))                                                  
> Out[6]: 
> x⋅(x - 1)
> ─────────
>     2  
> ```

Should we expand the binomial if we have a multivariable equation, example
 `binomial(x + y, 2)`

**Comment by oscarbenjamin** [comment]
> Should we expand the binomial if we have a multivariable equation

Yes if any of the symbols of interest is in the binomial. Otherwise we won't be able to solve the equations.

**Comment by sbt4104** [comment]
ok, I will make that change.

**Comment by sbt4104** [comment]
I have made the changes, please review.

**Comment by sbt4104** [comment]
> It looks like `expand_func` for binomial could be improved:
> 
> ```
> In [18]: expand_func(binomial(x, 2))                                                                                                                          
> Out[18]: 
> x⋅(x - 1)
> ─────────
>     2    
> 
> In [19]: expand_func(binomial(x+2, x))                                                                                                                        
> Out[19]: 
> ⎛x + 2⎞
> ⎜     ⎟
> ⎝  x  ⎠
> 
> In [20]: gammasimp(_.rewrite(factorial))                                                                                                                      
> Out[20]: 
> (x + 1)⋅(x + 2)
> ───────────────
>        2   
> ```

expand_func expands `binomial(x, x-1)` properly but fails for `binomial(x+1, x)`

I think changing 
`if k.is_Add and n in k.args : k = n - k` 
to
`if (k.is_Add and n in k.args) or (n.is_Add and k in n.args) : k = n - k`
will solve this problem

**Comment by oscarbenjamin** [comment]
Or maybe just `if (k - n).is_Integer`

**Comment by sbt4104** [comment]
yes, this one will also expand cases like
`solve_func(binomial(x + 1, x - 1))`

**Comment by sbt4104** [comment]
should I send a PR for this along with the issue?

**Comment by oscarbenjamin** [comment]
You can make those changes in this PR as well

**Comment by sbt4104** [comment]
I have made the changes, please review.

File: sympy/functions/combinatorial/factorials.py [hunk 3]
```diff
@@ -1012,7 +1012,7 @@ def _eval_expand_func(self, **hints):
             return binomial(*self.args)
 
         k = self.args[1]
-        if k.is_Add and n in k.args:
+        if (n-k).is_Integer:
             k = n - k
```
Comment (oscarbenjamin) [inline sympy/functions/combinatorial/factorials.py]: There should be a test for this.

File: sympy/solvers/solvers.py [hunk 4]
```diff
@@ -1460,6 +1460,9 @@ def _solve(f, *symbols, **flags):
             raise NotImplementedError(not_impl_msg % f)
     symbol = symbols[0]
 
+    #expand binomials only if it has the unknown symbol
+    f = f.replace(lambda e: isinstance(e, binomial) and e.has(symbol), lambda e: expand_func(e))
```
Comment (oscarbenjamin) [inline sympy/solvers/solvers.py]: This line is too long

File: sympy/solvers/tests/test_solvers.py [hunk 5]
```diff
@@ -2137,6 +2137,13 @@ def test_issue_17949():
     assert solve(exp(+x-x**2), x) == []
     assert solve(exp(-x-x**2), x) == []
 
+def test_issue_10993():
+    assert solve(Eq(binomial(x, 2), 3)) == [-2, 3]
+    assert solve(Eq(pow(x,2) + binomial(x, 3), x)) == [-4, 0, 1]
+    assert solve(Eq(binomial(x,2),0)) == [0, 1]
+    assert solve(a+binomial(x, 3),a) == [-binomial(x, 3)]
+    assert solve(x-binomial(a, 3) + binomial(y,2) + sin(a),x) == [-sin(a) + binomial(a, 3) - binomial(y, 2)]
+    assert solve((x+1)-binomial(x+1, 3),x) == [-2, -1, 3]
```
Comment (oscarbenjamin) [inline sympy/solvers/tests/test_solvers.py]: Include a space after a comma

**Comment by oscarbenjamin** [comment]
This looks good. I'll merge it once tests pass.

For future PRs can you try to use better commit messages? The messages here started out good but then got less good as time went on. You should imagine that the message will be looked at in future by someone who does not know the full context of the PR. Something like "add test" does not convey much about the change that is being made to the code.

**Comment by sbt4104** [comment]
I will keep it in mind, this will not happen again.

**Comment by oscarbenjamin** [comment]
Merged. Thanks for this!

**Comment by sbt4104** [comment]
Thank you, l learned a lot from this.

**Pull request merged**